mcproto/types/nbt.py

```diff
@@ -0,0 +1,1315 @@
+            If set to True, the tag will be read from the buffer and
+                the return type will be inferred from the tag type.
+            If set to False and called from a subclass, the tag type will be inferred from the subclass.
+            If set to False and called from the base class, the tag type will be TAG_Compound.
+
+            If with_type is set to False, the buffer must not start with the tag type byte.
+
+        :return: The deserialized NBT tag.
+        """
+        name, tag_type = cls._read_header(buf, with_name=with_name, read_type=with_type)
+
+        tag_class = NBTag.ASSOCIATED_TYPES[tag_type]
+        if cls not in (NBTag, tag_class):
+            raise TypeError(f"Expected a {cls.__name__} tag, but found a different tag ({tag_class.__name__}).")
+
+        tag = tag_class.read_from(buf, with_type=False, with_name=False)
+        tag.name = name
+        return tag
+
+    @classmethod
+    def _read_header(cls, buf: Buffer, read_type: bool = True, with_name: bool = True) -> tuple[str, NBTagType]:
+        """Read the header of the NBT tag.
+
+        :param buf: The buffer to read from.
+        :param read_type: Whether to read the type of the tag from the buffer.
+        :param with_name: Whether to read the name of the tag. If set to False, the tag will have the name "".
+
+        :return: A tuple containing the name and the tag type.
+
+
+        :note: It is possible that this function reads nothing from the buffer if both with_name and read_type are set
+            to False.
+        """
+        tag_type: NBTagType = cls.TYPE  # default value
+        if read_type:
+            try:
+                tag_type = NBTagType(buf.read_value(StructFormat.BYTE))
+            except OSError:
+                raise IOError("Buffer is empty.") from None
+            except ValueError:
+                raise TypeError("Invalid tag type.") from None
+
+        if tag_type == NBTagType.END:
+            return "", tag_type
+
+        name = StringNBT.read_from(buf, with_type=False, with_name=False).value if with_name else ""
+
+        return name, tag_type
+
+    @classmethod
+    def read_from(cls, buf: Buffer, with_type: bool = True, with_name: bool = True) -> NBTag:
+        """Read the NBT tag from the buffer.
+
+        :param buf: The buffer to read from.
+        :param with_type: Whether to read the type of the tag.
+            If set to True, the tag will be read from the buffer and
+                the return type will be inferred from the tag type.
+            If set to False and called from a subclass, the tag type will be inferred from the subclass.
+            If set to False and called from the base class, the tag type will be TAG_Compound.
+        :param with_name: Whether to read the name of the tag. If set to False, the tag will have the name "".
+
+        :return: The NBT tag.
+        """
+        return cls.deserialize(buf, with_name=with_name, with_type=with_type)
```

Comment:
This can't happen because NBTag can never be inside the `ASSOCIATED_TYPES` dict. I was not very familiar with abstract methods but this is a much better.

Reply:
Yeah it can't, but classes that inherit from it, and don't override this method can be in `ASSOCIATED_TYPES`, which means child classes have to override this. If that's the case, that means this should be an abstract method.

Reply:
I meant abstract methods is much better, that's what I converted it to :)